### DISCUSSION TITLE
Allow specifying `datetime` (and `time`) to be aware or naive

### DISCUSSION THREAD

**Discussion by shniubobo** [pydantic/pydantic#3477] [Feature Request]
## Add `AwareDatetime` and `NaiveDatetime` types

They should work like this:

```python
from datetime import datetime, timezone
from pydantic import BaseModel

class Naive(BaseModel):
    dt: NaiveDatetime

class Aware(BaseModel):
    dt: AwareDatetime

naive = Naive(dt=datetime.now(None))
aware = Aware(dt=datetime.now(timezone.utc))
```

## Coercion or error

What if, say, when an `AwareDatetime` field gets a naive `datetime`? Maybe we can further add `StrictAwareDatetime` and `StrictNaiveDatetime` which raise an exception when given a value of the wrong type. But if we decide to coerce non-strict types, how should we do so?

## Coercion rules

The main question is that, what should we treat naive `datetime` objects as? Do they represent local time or utc time? Since `datetime.astimezone` treats a naive instance as local time, ...

```python
>>> # Local time zone is UTC+08:00
>>> datetime(2021, 12, 3, 8).astimezone(timezone.utc)
datetime.datetime(2021, 12, 3, 0, 0, tzinfo=datetime.timezone.utc)
```

... I think we had better stick with that, and assume that the `datetime` should be converted to utc.

So the coercion should work like this:

```python
>>> coerced_aware = Aware(dt=datetime(2021, 12, 3, 8))
>>> coerced_aware.dt
datetime.datetime(2021, 12, 3, 0, 0, tzinfo=datetime.timezone.utc)
```

PS: If the feature request would be accepted, I'd like to open a PR to implement it.

**Comment by antonagestam** [comment]
Hi 👋

I maintain [phantom-types](https://github.com/antonagestam/phantom-types), a library for narrowing on builtin types. It comes with [`TZAware` and `TZNaive`](https://phantom-types.readthedocs.io/en/stable/pages/types.html#module-phantom.datetime) types out of the box and is compatible with pydantic. The `TZAware` type will never coerce an ambiguous value though, simply because it isn't safe.

In short, the difference between pydantic's constrained types and phantom types is compatibility with `isinstance()`.

It sounds like this would fulfill the requirements if you're willing to bring in a new dependency.

**Reply by samuelcolvin** [reply]
Thanks so much for this.

I think we should steal this idea from phantom-types and add tz aware/naive constraints to datetimes in pydantic-core core. Very happy to give credit to phantom-types for the idea.

@antonagestam any chance you'd be willing to get your hands dirty with some rust and make a PR??? 🙏

**Reply by antonagestam** [reply]
@samuelcolvin I'd really like to. I'll make an attempt to find the time, but can't make any promises at the moment.

**Reply by samuelcolvin** [reply]
Thanks so much. No pressure.

Let me know if you have questions https://github.com/pydantic/pydantic-core/issues/266

**Reply by tribals** [reply]
This is how "types with human face" could be.

**Reply by antonagestam** [reply]
@samuelcolvin Would a good next step here be to add `TZAware` and `TZNaive` to `pydantic.types`? Are `FutureDate`/`PastDate` good examples to look at?

**Reply by samuelcolvin** [reply]
Yea absolutely.

**Reply by antonagestam** [reply]
Cool, looking into it :)

**Comment by nameer** [comment]
Any update on this? Is there any plan?

**Reply by aiven-anton** [reply]
@nameer Samuel posted the plan above, it's here: https://github.com/pydantic/pydantic-core/issues/266. Feel free to dig in.

**Reply by antonagestam** [reply]
Here's an update on some progress:

- First step, adding support in pydantic-core, is merged: https://github.com/pydantic/pydantic-core/pull/343
- For the final step, I just pushed a candidate PR: https://github.com/pydantic/pydantic/pull/4812

Cheers!
